### PR TITLE
[Dart] Add configuration to add new generators and add json_serializable

### PR DESCRIPTION
**Pull request by agilob**
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by agilob** [comment]
@swipesight (2018/09) @jaumard (2018/09) @josh-burton (2019/12) @amondnet (2019/12) @sbu-WBT (2020/12) @kuhnroyal (2020/12) @agilob (2020/12)

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractDartCodegen.java [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+
+    public static final String PUB_LIBRARY = "pubLibrary";
+    public static final String PUB_NAME = "pubName";
+    public static final String PUB_VERSION = "pubVersion";
+    public static final String PUB_DESCRIPTION = "pubDescription";
+    public static final String PUB_AUTHOR = "pubAuthor";
+    public static final String PUB_AUTHOR_EMAIL = "pubAuthorEmail";
+    public static final String PUB_HOMEPAGE = "pubHomepage";
+    public static final String USE_ENUM_EXTENSION = "useEnumExtension";
+    public static final String SERIALIZATION_LIBRARY_CUSTOM = "custom";
```
Comment (kuhnroyal) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractDartCodegen.java]: I would call this native

Comment (kuhnroyal) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractDartCodegen.java]: And I would expect these in the `DartClientCodegen`

**Review by kuhnroyal — COMMENTED** [review]
I like it a lot.
You should test if the minimum supported Dart version needs to be raised.
And maybe get rid of the new empty lines to reduce noise.

Also add a pom.xml for the json_serializable sample?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractDartCodegen.java (hunk 1 above)
Comment (agilob) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractDartCodegen.java]: >I would call this native

Thanks, I had hard time coming up with good name.

>And I would expect these in the DartClientCodegen

Diff is outdated, it is there now!

**Comment by agilob** [comment]
> You should test if the minimum supported Dart version needs to be raised.

json_serializer now has version 4.0.0 which **will** require updating dart version, but so far this is ok. I don't want to make too many changes at once here. eg. default value isn't support yet here, nor standalone enums, but it's a major move forward I think. It will open way for @k-paxian to add dart-json-mapper  serializer in client library.

**Comment by kuhnroyal** [comment]
I am not the current approach with the part files will work. Have you run the build runner? I think you are missing the model part files and this since only one `part (of)` is allowed, this might fail.

Maybe we first need to get rid of the current part files and use correct exports.

**Comment by agilob** [comment]
>I am sure not the current approach with the part files will work. Have you run the build runner? I think you are missing the model part files and since only one part (of) is allowed, this might fail.

Neither I am, but this actually works quite nice. Instead producing really verbose files, one .g.dart for each model it generates one big `api.g.dart` which has serialization code for all models. It's much nicer to see in IDE. The project builds:

```
❯ pwd
/home/agilob/Projects/openapi-generator/samples/openapi3/client/petstore/dart2/petstore_json_serializable_client_lib_fake
 pub get && pub run build_runner build
Resolving dependencies... 
Got dependencies!
Precompiling executable... (8.6s)
Precompiled build_runner:build_runner.
[INFO] Generating build script completed, took 242ms
[INFO] Reading cached asset graph completed, took 44ms
[INFO] Checking for updates since last build completed, took 324ms
[INFO] Running build completed, took 19ms
[INFO] Caching finalized dependency graph completed, took 25ms
[INFO] Succeeded after 51ms with 0 outputs (0 actions)


```

**Comment by agilob** [comment]
> And maybe get rid of the new empty lines to reduce noise.

Check commit history... and I give up, the new empty lines are result of
```
{{/json_serializable}}
{{#json_serializable}}
```

I'd rather have nicer mustache template than autogenerated code that shouldn't me modified manually anyway... (and we recommend running dartfmt after generation). You can [remove whitespaces](https://github.com/OpenAPITools/openapi-generator/pull/8789/files?diff=split&w=1) from GH review in options.

**Comment by agilob** [comment]
Travis failing
```
anyrequests: You have reached your pull rate limit.
You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

@wing328

**Comment by kuhnroyal** [comment]
That is good but I know a lot of generators where this will not work. So we should do that anyway, but then it can be independent of this change.

There seems to be some transformer stuff left in the api client and helper that doesnt compile.

**Comment by agilob** [comment]
>That is good but I know a lot of generators where this will not work. So we should do that anyway, but then it can be independent of this change.

I propose to do it after this change, I find this PR to be too much in one go already.

**Comment by agilob** [comment]
>There seems to be some transformer stuff left in the api client and helper that doesnt compile.

Unfortunately this is where we're hitting temporary limitation of json_serializable and for now I can't add this module to pom.xml

image

it doesn't generate mapping for enums that are not used in **classes**. This is purely temporary as this api-helper and api-client are redundant right now. Each class has its own toJson and fromJson method now, but I'm going to keep it in this PR for backwards consistency. Both native and j_s serializes now can be used interchangeably.

**Comment by kuhnroyal** [comment]
I see, so you want this merged to have the basic infrastructure in place although the json_serializable option does not completely work yet. I agree in that regard, maybe we should log some warning when this option is set until it is completely working. 

> > That is good but I know a lot of generators where this will not work. So we should do that anyway, but then it can be independent of this change.
> 
> I propose to do it after this change, I find this PR to be too much in one go already.

I agree, this is out of scope.

**Comment by agilob** [comment]
> I agree in that regard, maybe we should log some warning when this option is set until it is completely working.

I'll add default value support after this gets in. It doesn't have to be perfect on the first merge. default value is also incompatible with `nullable` flag in json_serializable so this need to be conditional support in mustache template, which I simply avoided doing now. This is big already.

If you think this looks good, can we get this merged please @kuhnroyal  and @wing328 ?

**Comment by kuhnroyal** [comment]
I think changing the enum templates to `}{{/json_serializable}}{{#json_serializable}}` removes the new lines if you want. Otherwise I am good with this.

**Comment by agilob** [comment]
Done!

**Review by kuhnroyal — APPROVED** [review]

**Comment by wing328** [comment]
@agilob thanks for the PR. Please resolve the merge conflicts when you've time.

**Comment by agilob** [comment]
@wing328 done

**Comment by agilob** [comment]
@wing328 can we get this in please? I want to add a few more features to this generator while I still have time for it

**Comment by wing328** [comment]
Just merged. Thanks again for the PR.

**Pull request merged**